### PR TITLE
CVS-46712 make pipeline factory spawn pipeline with demultiplexer and gather nodes

### PR DESCRIPTION
**Pull request by Ficol**

### CODE REVIEW

File: src/node.cpp [hunk 1]
```diff
@@ -27,9 +27,10 @@
 
 namespace ovms {
 
-Node::Node(const std::string& nodeName, uint32_t demultiplyCount) :
+Node::Node(const std::string& nodeName, uint32_t demultiplyCount, std::set<std::string> gatherFromNode) :
     nodeName(nodeName),
-    demultiplexCount(demultiplyCount ? std::optional<uint32_t>(demultiplyCount) : std::nullopt) {
+    demultiplexCount(demultiplyCount ? std::optional<uint32_t>(demultiplyCount) : std::nullopt),
+    gatherFrom(!gatherFromNode.empty() ? std::optional<std::set<std::string>>(gatherFromNode) : std::nullopt) {
```
Comment (dkalinowski) [inline src/node.cpp:33]: Do we need to use optional of set of strings? Maybe we could rely on a fact that empty set means no gathering at all.

File: src/pipelinedefinition.cpp [hunk 2]
```diff
@@ -151,9 +151,18 @@ Status PipelineDefinition::create(std::unique_ptr<Pipeline>& pipeline,
     std::unordered_map<std::string, std::unique_ptr<Node>> nodes;
     EntryNode* entry = nullptr;
     ExitNode* exit = nullptr;
+    std::set<std::string> gatherFrom;
```
Comment (dkalinowski) [inline src/pipelinedefinition.cpp:154]: Maybe demultiplyNodeNames or demultipliers would be better name?

File: src/pipelinedefinition.cpp [hunk 3]
```diff
@@ -151,9 +151,18 @@ Status PipelineDefinition::create(std::unique_ptr<Pipeline>& pipeline,
...
+    for (const auto& info : nodeInfos) {
+        if (info.demultiplyCount) {
+            gatherFrom.insert(info.nodeName);
+        }
+    }
     for (const auto& info : nodeInfos) {
         SPDLOG_LOGGER_DEBUG(dag_executor_logger, "Creating pipeline: {}. Adding nodeName: {}, modelName: {}",
             getName(), info.nodeName, info.modelName);
+        if (info.gatherFromNode) {
+            gatherFrom.erase(info.nodeName);
```
Comment (dkalinowski) [inline src/pipelinedefinition.cpp:164]: We should check if erase did in fact erased an element. If not, this means pipeline is misconfigured and we shouldn't proceed with further execution

File: src/node.cpp (hunk 1 above)
Comment (atobiszei) [reply src/node.cpp:33]: While this suggestion makes sens I think this is out of scope of this MR.

File: src/pipelinedefinition.cpp (hunk 3 above)
Comment (atobiszei) [reply src/pipelinedefinition.cpp]: You could gather on separate branches of DAG before going into exit node.

**Review by atobiszei — APPROVED** [review]

**Review by dkalinowski — APPROVED** [review]

**Pull request merged**